### PR TITLE
feat: Add possibility to specify write-back GIT repository as annotation

### PR DESCRIPTION
**Pull request by flozzone**
Addressing #423

This PR does the following:
* Get the repository URL either from `app.Spec.Source.RepoURL` or from `argocd-image-updater.argoproj.io/git-repository` if specified and store it into the `WriteBackConfig` struct as `GitRepo`.
* Write a test to show `git-repository` annotation functionality
* Update relevant documentation
* Fix GitHub workflow linting tests by upgrading `golangci-lint` to its latest version

### CODE REVIEW

**Comment by flozzone** [comment]
Do we need to update something in order to pass linting tests? #413 #410 seem to have a similar problem.

**Comment by flozzone** [comment]
> Do we need to update something in order to pass linting tests? #413 #410 seem to have a similar problem.

Updating to latest version `v1.45.2` for the `golangci/golangci-lint-action@v2` action fixed linting tests.

**Comment by nikodemjedynak-dnv** [comment]
@flozzone Any updates on this resolution to the issue? Is it going to be merged? Anyone fancies a review?

**Review by chathumal-nad — APPROVED** [review]

File: .github/workflows/ci-tests.yaml [hunk 1]
```diff
@@ -57,7 +57,7 @@ jobs:
       - name: Run golangci-lint
         uses: golangci/golangci-lint-action@v2
         with:
-          version: v1.32
+          version: v1.45.2
```
Comment (jlifts) [inline .github/workflows/ci-tests.yaml:60]: This is the only merge conflict I see and when accepting the incoming changes it passes

**Review by jlifts — COMMENTED** [review]
I think this is solid work, the only merge conflict is the ci-tests golangci-lint-action version. I tested it locally and the tests still pass. I would approve but not sure what the rules around that look like (if a more frequent contributer is required to approve).

**Review by jlifts — APPROVED** [review]

File: .github/workflows/ci-tests.yaml (hunk 1 above)
Comment (flozzone) [reply .github/workflows/ci-tests.yaml]: I needed to upgrade to golangci-lint to make the tests pass. In the meanwhile also master branch has updated golangci-lint to v1.46.2 so I've removed this commit.

Comment (flozzone) [reply .github/workflows/ci-tests.yaml]: Still, linting fails in CI workflow with:

```
Running [/home/runner/golangci-lint-1.46.2-linux-amd64/golangci-lint run --out-format=github-actions --timeout 5m] in [] ...
  panic: load embedded ruleguard rules: rules/rules.go:13: can't load fmt
  
  goroutine 1 [running]:
  github.com/go-critic/go-critic/checkers.init.22()
  	github.com/go-critic/go-critic@v0.6.3/checkers/embedded_rules.go:47 +0x4b4
  
  Error: golangci-lint exit with code 2
  Ran golangci-lint in 9[72](https://github.com/argoproj-labs/argocd-image-updater/actions/runs/4968280408/jobs/8890784534?pr=424#step:3:74)2ms
```
https://github.com/argoproj-labs/argocd-image-updater/actions/runs/4968280408/jobs/8890784534?pr=424

Going to update `golangci-lint` to most recent version.

Comment (flozzone) [reply .github/workflows/ci-tests.yaml]: To fix `golangci-lint` issues I needed to:
* Update `golangci-lint` to 1.52.2
* Replace deprecated linters `deadcode`, `varcheck`, `structcheck` by the newer `unused` linter. See relevant `golangci-lint` PR at https://github.com/golangci/golangci-lint/issues/1841
* Run `goimports -w -local github.com/argoproj-labs/argocd-image-updater .` to fix
** Comment indentation
** Reorder imports

**Comment by flozzone** [comment]
To fix golangci-lint issues I needed to:

Update golangci-lint to 1.52.2
* Replace deprecated `linters`, `deadcode`, ,`varcheck`, `structcheck` by the newer `unused` linter. See relevant golangci-lint PR at https://github.com/golangci/golangci-lint/issues/1841
* Run goimports -w -local github.com/argoproj-labs/argocd-image-updater . to fix
  * Comment indentation
  * Reorder imports

**Comment by jannfis** [comment]
Thanks for bumping this @bn-cforkner , I'm currently going through the backlog of PRs and issues. I'm really sorry to have been undercover for a while.

Couple of the changes in this PR have been merged recently (i.e. lint update), @flozzone could you please rebase this PR to what's in master? Thanks!

**Comment by rwc** [comment]
@jannfis @flozzone Hoping to get another look before this gets anymore stale. Would definitely provide some useful function we’re excited to get our hands on.

**Comment by LS80** [comment]
@flozzone I'm interested in the use case for this. As far as I can tell, the override files would not be used by Argo CD when rendering the manifests so why write the files to a separate git repository?

**Comment by jlifts** [comment]
> @flozzone I'm interested in the use case for this. As far as I can tell, the override files would not be used by Argo CD when rendering the manifests so why write the files to a separate git repository?

@LS80 I know a use case it could be handy for is when you want git write back method and are using an OCI storage for your charts.

**Comment by LS80** [comment]
> @LS80 I know a use case it could be handy for is when you want git write back method and are using an OCI storage for your charts.

Yes but I'm not sure what use the git write back method has if the override files created/updated in git will not actually be used by Argo CD. If an Argo CD application could be configured to use override files from a repo other than the one specified as `source.repoURL` then it would be really useful. I've just remembered that we can now have multiple `sources` so maybe that would work. One repoURL for the parameter override files in git and another referencing the helm repo...

**Comment by rwc** [comment]
@LS80 We use it as a record of deployment in environments where the origin of what’s running must be certified/tracked for compliance reasons.

**Pull request merged**